src/backend/executor/exchange_hash_join_executor.cpp

```diff
@@ -0,0 +1,365 @@
+        }
+      }
+    }
+    // Check if we have any join tuples
+    if (pos_lists_builder.Size() > 0) {
+      LOG_TRACE("Join tile size : %lu \n", pos_lists_builder.Size());
+      output_tile->SetPositionListsAndVisibility(pos_lists_builder.Release());
+      lockfree_buffered_output_tiles.push(output_tile.release());
+    }
+  }
+
+  LOG_TRACE("Probe() thread %u done", (unsigned)self_no);
+  barrier->Release();
+}
+
+/**
+ * @brief Creates logical tiles from the two input logical tiles after applying
+ * join predicate.
+ * @return true on success, false otherwise.
+ */
+bool ExchangeHashJoinExecutor::DExecute() {
+  // Loop until we have non-empty result tile or exit
+  for (;;) {
+    if (lockfree_buffered_output_tiles.empty() == false) {
+      LogicalTile *output_tile = nullptr;
+      //          bool ret = lockfree_buffered_output_tiles.pop(output_tile);
+      //          assert(ret);
+      lockfree_buffered_output_tiles.pop(output_tile);
+      SetOutput(output_tile);
+      // exit 0
+      return true;
+    }
+
+    if (prepare_children_ == false) {
+      // build right hashTable
+      Barrier build_hashtable_barrier(1);
+      std::function<void()> build_hashtable_worker =
+          std::bind(&ExchangeHashJoinExecutor::GetRightHashTable, this,
+                    &build_hashtable_barrier);
+      LaunchWorkerThreads(1, build_hashtable_worker);
+      LOG_INFO("Wait for right child build to finish.\n");
+
+      // collect all left children
+      Barrier collect_scan_result_barrier(1);
+      std::function<void()> collect_scan_result_worker =
+          std::bind(&ExchangeHashJoinExecutor::GetLeftScanResult, this,
+                    &collect_scan_result_barrier);
+      LaunchWorkerThreads(1, collect_scan_result_worker);
+      LOG_INFO("Wait for left child scan to finish.\n");
+
+      build_hashtable_barrier.Wait();
+      collect_scan_result_barrier.Wait();
+      LOG_INFO("Ready to Probe.\n");
+
+      if ((right_result_tiles_.size() == 0) &&
+          (join_type_ == JOIN_TYPE_INNER || join_type_ == JOIN_TYPE_RIGHT)) {
+        no_need_to_probe_ = true;
+
+        return false;
+      } else if ((left_result_tiles_.size() == 0) &&
+                 (join_type_ == JOIN_TYPE_INNER ||
+                  join_type_ == JOIN_TYPE_LEFT)) {
+        no_need_to_probe_ = true;
+        return false;
+      }
+
+      main_start = std::chrono::system_clock::now();
+
+      // partition sub tasks
+      size_t left_child_size = left_result_tiles_.size();
+      if (left_child_size == 0) {
+        probe_barrier_.SetNeedToDo(true);
+      }
+      size_t partition_number = left_child_size / SIZE_PER_PARTITION;
+      if (left_child_size % SIZE_PER_PARTITION != 0) {
+        ++partition_number;
+      }
+      LOG_INFO("left_result_tiles.size():%lu, partition num in probe:%lu\n",
+               left_child_size, partition_number);
+
+      // sub tasks begin
+      std::atomic<thread_no> no(0);
```

Comment:
where do we join the threads launched here? some comments would help with readability. also, "no" should have a more descriptive variable name. Does it fall out of scope at the end of this function?